### DISCUSSION TITLE
How to persist state using Pinia + Quasar LocalStorage Plugin?

### DISCUSSION THREAD

**Discussion by BenJackGill** [quasarframework/quasar#12376] [General - Components / Directives / etc] [answered]
I'm using Pinia for state managment, and I want the state to persist when the page is refeshed.

I'm aware of two options:

1) Use a plugin. Vuex has a [vuex-persistedstate](https://github.com/robinvdvleuten/vuex-persistedstate) plugin for this, and Pinia has [a similar plugin](https://github.com/Seb-L/pinia-plugin-persist) but it's still under development.

2) Use local storage. Luckily Quasar has a LocalStorage plugin which would be nice to use here. But I'm not sure how to integrate it with Pinia, thus the reason for this post.

I found a nice [tutorial](https://stephanlangeveld.medium.com/simple-local-storage-implementation-using-vue-3-vueuse-and-pinia-with-zero-extra-lines-of-code-cb9ed2cce42a) doing something similar with Pinia + [Vueuse](https://vueuse.org/).

I tried adapting it to my needs with Pinia + TypeScript + Quasar LocalStorage Plugin as per below:

```
import { User } from 'firebase/auth';
import { defineStore } from 'pinia';
import { LocalStorage } from 'quasar';

type CreateMutable<T> = { -readonly [P in keyof T]: CreateMutable<T[P]> };

export const useUserStore = defineStore('user', {
  state: () => ({
    // user: {} as CreateMutable<User>, // This was my previous non-persistent state
    user: LocalStorage.set('user', {}) as unknown as CreateMutable<User>, // This is my attempt at using LocalStorage persistent state
  }),
  actions: {
    setUser(userData: User) {
      this.user = userData;
    },
    setPhotoURL(photoURLData: string | null) {
      this.user.photoURL = photoURLData;
    },
  },
});
```
Unfortunately it doesn't work.

The state does not persist on page refresh.

Any ideas on how to make this work?

**Comment by yusufkandemir** [comment, marked as answer]
Hi, `LocalStorage.set` doesn't return anything, so `user: LocalStorage.set('user', {}) as unknown as CreateMutable<User>,` is essentially `user: undefined`. You must use `LocalStorage.getItem()` to get the initial value, then use `LocalStorage.set()` when the value needs to be updated.

Here is an example of just saving the state to local storage, both scoped to a single variable, and to the global store:
https://pinia.vuejs.org/core-concepts/state.html#subscribing-to-the-state
Here is the adaptation:
```ts
import { User } from 'firebase/auth';
import { defineStore } from 'pinia';
import { LocalStorage } from 'quasar';

type CreateMutable<T> = { -readonly [P in keyof T]: CreateMutable<T[P]> };

export const useUserStore = defineStore('user', {
  state: () => ({
    // user: {} as CreateMutable<User>,
    user: LocalStorage.getItem<CreateMutable<User>>('user') || {} as CreateMutable<User>,
  }),
  actions: {
    setUser(userData: User) {
      this.user = userData;
    },
    setPhotoURL(photoURLData: string | null) {
      this.user.photoURL = photoURLData;
    },
  },
});

// Then, inside App.vue's setup()
const userStore = useUserStore();

userStore.$subscribe((mutation, state) => {
  LocalStorage.set('user', state.user)
})
```

Alternatively, you can create a Pinia plugin just for yourself. These should help:
https://pinia.vuejs.org/core-concepts/plugins.html
https://pinia.vuejs.org/core-concepts/plugins.html#calling-subscribe-inside-plugins
https://pinia.vuejs.org/core-concepts/plugins.html#adding-new-options

**Reply by BenJackGill** [reply]
Works great!

Only thing is there was a small typo. It should be `LocalStorage.getItem<CreateMutable<User>>('user')` not `LocalStorage.get<CreateMutable<User>>('user')`. Just FYI incase anyone else reads this later and wants to use it for their app.

**Reply by yusufkandemir** [reply]
You are welcome! I updated my answer to make it easier. Thanks for pointing that out.
